### DISCUSSION TITLE
User Migration from external Service

### DISCUSSION THREAD

**Discussion by Bhavik237** [keycloak/keycloak#21062] [Q&A] [answered]
Greetings, everyone,
I am in the process of transitioning users from an external service to Keycloak. During this migration, I will transfer all user metadata from the external service to Keycloak, with the exception of the password, which is stored in a hashed format.

Consequently, users will be present in Keycloak without their credentials. They will need to reset or set their password when they enter the Keycloak system. Here's how it will work:
![User migration (1)](https://github.com/keycloak/keycloak/assets/57186072/2e182e49-01fd-46ec-a5f1-89463f9c0a2b)

(Email is compulsory)
Users' usernames will be available in the system without associated credentials. When they enter their username in Keycloak, they will receive an email allowing them to reset their password.

If this approach is feasible, I would appreciate guidance on the recommended steps to implement it.

Thanks.

**Comment by vplemyannik** [comment, marked as answer]
I dont know how to configure your flow, but i can point to other options.

1.  If you know how the hash algorithm is implemented in external services, then you sync hashed passwords into keycloak, after that you can create your own plugin with implemented PasswordHashProviderFactory here is an example https://github.com/pavelbogomolenko/keycloak-custom-password-hash/blob/master/src/main/java/org/keycloak/examples/providers/credential/hash/CustomPasswordHashProvider.java

2. (How i did it) implemente UserStorageProvider,  ImportSynchronization, CredentialInputValidator. 
UserStorageProvider and ImportSynchronization are responsible for sync users between external service and keycloak. But  with CredentialInputValidator you can use to delegate authentication process to external system and if it is success then save password in keycloak
```java
 @Override
    public boolean isValid(RealmModel realmModel, UserModel userModel, CredentialInput credentialInput) {
            //  credentialInput.getChallengeResponse() raw password 
            var result= externalService.login(userModel.getUsername(), credentialInput.getChallengeResponse());
            if(result.getStatus() ==LoginStatus.FAILED){
                return false;
            }
            // save password to keycloak
            userModel.credentialManager().updateCredential(credentialInput);
    }

```
when all users will be login at least once  you can disable your custom user federation 
here is an example 
https://github.com/RiskManagementGroup/keycloak-msg-api-user-federation/blob/main/src/main/java/dk/rmgroup/keycloak/storage/api/msg/MsgApiUserStorageProviderFactory.java

**Reply by Bhavik237** [reply]
No I don't know hash implementation of external system.  
But i am using this flow 
![Screenshot from 2023-06-19 16-47-20](https://github.com/keycloak/keycloak/assets/57186072/abb4701a-6cb4-4204-9a23-e20b86863ec9)

while importing user I will set attribute (name = toBeRest value= yes)  after login I will change attribute value

This flow is working but I have one problem after sending reset password mail , I am getting user form screen (idle I should password form)
